reference-implementation/lib/helpers.js

```diff
@@ -157,11 +157,38 @@ exports.MakeSizeAlgorithmFromSizeFunction = size => {
   return chunk => size(chunk);
 };
 
-exports.PerformPromiseThen = (promise, onFulfilled, onRejected) => {
+const originalPromise = Promise;
+const originalPromiseThen = Promise.prototype.then;
+const originalPromiseResolve = Promise.resolve;
+const originalPromiseReject = Promise.reject;
+
+function CreatePromise(executor) {
+  return new originalPromise(executor);
+}
+
+function PromiseResolve(value) {
+  return originalPromiseResolve.call(originalPromise, value);
+}
+
+function PromiseReject(reason) {
```

Comment:
I wasn't sure what to name these functions, since there's no such thing as `PromiseReject` or `PerformPromiseCatch` in [the ECMAScript specification](https://tc39.es/ecma262/). I considered naming them after the phrases in the promise guide, but I found `PromiseRejectedWith(reason)` and `UponRejection(promise, onRejected)` a bit weird.

Feel free to suggest better names. 😉

Reply:
Generally where we have a function that's not an abstract operation we start it with a lowercase letter, but I'm not 100% sure whether that's the right thing to do with these.

`PerformPromiseCatch` is probably too misleading, as it makes it sound like it is a standard operation. I think we can split it into two cases: `setPromiseIsHandled()` and `rethrowAssertions()`.

I think I prefer to use the names from the promises guide, so we have `aNewPromise()`, `aPromiseRejectedWith()`, and `aPromiseResolvedWith()`. Yes, these names are ugly but they help the reference implementation match the standard text. We should probably explicitly write `aPromiseResolvedWith(undefined)` to get it to read smoothly.

Since this contradicts what you want to do, I'd like to ask @domenic to be the stylistic judge here.

Reply:
I agree with sticking with the promises guide names, and using lower case. After all, this is the reference implementation; the closer to the standard text it can be, the better.

Reply:
Sounds reasonable. I'll rename them. 🙂